### ISSUE TITLE
Unexpected undefined-variable error in lambda functions

### ISSUE THREAD

**Issue by davidjlloyd** [PyCQA/pylint#1824]
### Steps to reproduce

Create a file `test.py` and add the following to it.

```python
class MyObject(object):
	method1 = lambda function: function()
	method2 = lambda func: func()
```

Run pylint against the file.

### Current behavior

The variable name `function` is considered undefined in `method1` but `func` in `method2` is fine! Pylint produces the following output:

```
E:  2,28: Undefined variable 'function' (undefined-variable)
R:  1, 0: Too few public methods (0/1) (too-few-public-methods)
```

### Expected behavior

No undefined variable errors.

### pylint --version output

pylint 1.7.1, 
astroid 1.5.3
Python 2.7.12 (default, Nov 20 2017, 18:23:56) 
[GCC 5.4.0 20160609]

**Comment by PCManticore** [comment]
Thanks, I can reproduce  this.

**Comment by MikhailArkhipov** [comment]
See also https://github.com/Microsoft/vscode-python/issues/891

**Comment by brycepg** [comment]
Thanks for the report. You can find the fix on both `master` and the `1.8` branch.

**Comment by AlJohri** [comment]
@brycepg I seem to be still getting this issue with the following code snippet:

```python
def compose(*functions):
    return reduce(lambda f, g: lambda x: f(g(x)), functions, lambda x: x)
```

pylint error:
```
Undefined variable: 'f'
Undefined variable: 'g'
```

```
$ pylint --version
pylint 2.3.1
astroid 2.2.5
Python 3.7.3 (default, May  7 2019, 13:48:34) 
[Clang 10.0.1 (clang-1001.0.46.4)]
```

If needed, example usage of above function:
```python
normalize_section = compose(
    fix_double_section, remove_double_slash, ensure_leading_slash, remove_trailing_slash
)
```
